UserExperience_Preload_Requests_Extension.php

```diff
@@ -0,0 +1,202 @@
+			return;
+		}
+
+		// Applies logic to display page cache flush notice if Preload Requests settings are altered and saved.
+		add_filter( 'w3tc_save_options', array( $this, 'w3tc_save_options' ) );
+
+		// Renders the Preload Reqeusts settings metabox on the User Expereince advanced setting page.
+		add_action( 'w3tc_userexperience_page', array( $this, 'w3tc_userexperience_page' ) );
+
+		// This filter is documented in Generic_AdminActions_Default.php under the read_request method.
+		add_filter( 'w3tc_config_key_descriptor', array( $this, 'w3tc_config_key_descriptor' ), 10, 2 );
+
+		// Applies dns-prefetch, preconnect, and preload headers.
+		add_action( 'wp_head', array( $this, 'w3tc_preload_requests_headers' ) );
+		add_action( 'admin_head', array( $this, 'w3tc_preload_requests_headers' ) );
+	}
+
+	/**
+	 * Renders the user experience Preload Requests settings page.
+	 *
+	 * @since 2.5.1
+	 *
+	 * @return void
+	 */
+	public function w3tc_userexperience_page() {
+		include __DIR__ . '/UserExperience_Preload_Requests_Page_View.php';
+	}
+
+	/**
+	 * Specify config key typing for fields that need it.
+	 *
+	 * @since 2.5.1
+	 *
+	 * @param mixed $descriptor Descriptor.
+	 * @param mixed $key Compound key array.
+	 *
+	 * @return array
+	 */
+	public function w3tc_config_key_descriptor( $descriptor, $key ) {
+		if (
+			is_array( $key )
+			&& in_array(
+				implode( '.', $key ),
+				array(
+					'user-experience-preload-requests.dns-prefetch',
+					'user-experience-preload-requests.preconnect',
+					'user-experience-preload-requests.preload-css',
+					'user-experience-preload-requests.preload-js',
+					'user-experience-preload-requests.preload-fonts',
+					'user-experience-preload-requests.preload-images',
+					'user-experience-preload-requests.preload-videos',
+					'user-experience-preload-requests.preload-audio',
+					'user-experience-preload-requests.preload-documents',
+				)
+			)
+		) {
+			$descriptor = array( 'type' => 'array' );
+		}
+
+		return $descriptor;
+	}
+
+	/**
+	 * Performs actions on save.
+	 *
+	 * @since 2.5.1
+	 *
+	 * @param array $data Array of save data.
+	 *
+	 * @return array
+	 */
+	public function w3tc_save_options( $data ) {
+		$new_config = $data['new_config'];
+		$old_config = $data['old_config'];
+
+		$new_includes = $new_config->get_array( array( 'user-experience-preload-requests', 'includes' ) );
+		$old_includes = $old_config->get_array( array( 'user-experience-preload-requests', 'includes' ) );
+
+		if ( $new_includes !== $old_includes && $this->config->get_boolean( 'pgcache.enabled' ) ) {
+			$state = Dispatcher::config_state();
+			$state->set( 'common.show_note.flush_posts_needed', true );
+			$state->save();
+		}
+
+		return $data;
+	}
+
+	/**
+	 * Applies the Preload Requests headers for wp_head and admin_head.
+	 *
+	 * @since 2.5.1
+	 *
+	 * @return void
+	 */
+	public function w3tc_preload_requests_headers() {
```

Comment:
All of the [esc_url()](https://developer.wordpress.org/reference/functions/esc_url/#more-information) calls will prepend "http://" to all URLs without a schema.  I used the suggested format "//fonts.googleapis.com" and I got mixed content errors with browsing an HTTPS page.

Reply:
Due to the fact that the protocol would be unknown for entries omitting it, there would be no way to determine the correct protocol. As a result I left the esc_url as is and updated the descriptions to indicate that values omitting the protocol would default to non-HTTPS